### PR TITLE
Add RISCV purecap system call assembly macros

### PR DESCRIPTION
**Pull request by arichardson**

### CODE REVIEW

**Comment by jrtc27** [comment]
Not sure why upstream isn't using `call`/`tail` for these and is instead using function pointers, but the purecap side of things looks correct.

**Comment by arichardson** [comment]
Does call allow interposition?

File: lib/libc/riscv/SYS.h [hunk 1]
```diff
@@ -42,6 +42,24 @@
...
+#else
+#define _GET_FNPTR(outreg, function)	clgc outreg, _C_LABEL(function)
+#define _CALL_FNPTR(fnptr)	cjalr cra, fnptr
+#define _TAILCALL_FNPTR(fnptr)	cjr fnptr
+#define _CALL_TMPREG	ct1
+#endif
+#define ASM_TAILCALL(function)					\
+	_GET_FNPTR(_CALL_TMPREG, function);			\
+	_TAILCALL_FNPTR(_CALL_TMPREG)
+#define ASM_CALL(function)					\
```
Comment (bsdjhb) [inline lib/libc/riscv/SYS.h]: ASM_CALL doesn't seem to be used?

Note that you could permit 't1' to be specified as a parameter and use `c ## reg` in the purecap version of the macro to convert it to a cap register.

**Comment by bsdjhb** [comment]
Pretty sure `cerror` is a local symbol that isn't meant to be preempted.  It sets `errno` and then returns to the original caller and is a stub rather than a full function.  The MIPS version just uses `j` for purecap.  I feel like this might actually be something that could use `cllc` rather than a GOT indirection perhaps, though if you had per-syscall stub bounds on pcc you would have to use `clgc`.

File: lib/libc/riscv/SYS.h (hunk 1 above)
Comment (arichardson) [reply lib/libc/riscv/SYS.h]: Yes that sounds good.
I added the CALL version since I think I'll need it in RTLD.

File: lib/libc/riscv/SYS.h [hunk 2]
```diff
@@ -56,8 +74,7 @@ ENTRY(__sys_##name);						\
 	_SYSCALL(name);						\
 	bnez	t0, 1f; 					\
 	ret;							\
```
Comment (arichardson) [inline lib/libc/riscv/SYS.h]: This needs to be cret, so I need to add another macro.

File: lib/libc/riscv/SYS.h [hunk 3]
```diff
@@ -42,22 +42,46 @@
 	li	t0, SYS_ ## name;				\
 	ecall
 
+#ifndef __CHERI_PURE_CAPABILITY__
+#define _GET_LOCAL_FNPTR(outreg, function)	\
+	lla outreg, _C_LABEL(function)
+#define _GET_FNPTR(outreg, function)	\
+	la outreg, _C_LABEL(function)
+#define _CALL_FNPTR(reg)	jalr cra, reg
```
Comment (bsdjhb) [inline lib/libc/riscv/SYS.h]: s/cra/ra/?  You can also just use `jalr reg` as a pseudo-instruction where `ra` is implied.

File: lib/libc/riscv/SYS.h [hunk 4]
```diff
@@ -42,22 +42,46 @@
...
+#define _GET_FNPTR(outreg, function)	\
+	clgc CAPABILITY_REG(outreg), _C_LABEL(function)
+#define _GET_LOCAL_FNPTR(outreg, function)	\
+	cllc CAPABILITY_REG(outreg), _C_LABEL(function)
+#define _CALL_FNPTR(reg)	cjalr cra, CAPABILITY_REG(reg)
+#define _TAILCALL_FNPTR(reg)	cjr CAPABILITY_REG(reg)
+#define _RETURN	cret
+#endif
+
+#define ASM_LOCAL_TAILCALL(function)				\
```
Comment (bsdjhb) [inline lib/libc/riscv/SYS.h]: I would probably pass the register to these so it's clearer to the reader which register is being used as a temporary.

File: lib/libc/riscv/SYS.h [hunk 5]
```diff
@@ -42,22 +42,46 @@
...
+	_GET_LOCAL_FNPTR(t1, function);				\
+	_TAILCALL_FNPTR(t1)
+#define ASM_LOCAL_CALL(function)				\
+	_GET_LOCAL_FNPTR(t1, function);				\
+	_TAILCALL_FNPTR(t1)
```
Comment (bsdjhb) [inline lib/libc/riscv/SYS.h]: s/TAIL//?

File: lib/libc/riscv/SYS.h [hunk 6]
```diff
@@ -42,22 +42,46 @@
 	li	t0, SYS_ ## name;				\
 	ecall
 
+#ifndef __CHERI_PURE_CAPABILITY__
```
Comment (bsdjhb) [inline lib/libc/riscv/SYS.h]: Slight preference for using #ifdef and putting purecap case first as you did in <machine/asm.h>

**Pull request merged**